fix for #806

pratikpparikh opened the pull request:
@kevmoo This is a fix for  #806.  Please review it and let me know if you find gaps.

kevmoo commented:
@pratikpparikh – check the red here.

Analyzer should be green.

I think the tests will fail w/ this, too!

You cannot cast a JSON value to `List<String>`. The type is `List<dynamic>` or `List<Object?>`.

kevmoo commented:
Didn't mean to close.

pratikpparikh commented:
@kevmoo still working on it.  I realized making generics work is breaking generics_test and kitchen_sink.  Trying to fix it now, maybe I will require some help.

pratikpparikh commented:
@kevmoo can you please suggest how i address analyzer_and_format issues?  All the test cases now work and I also fixed may gaps today.

kevmoo commented:
Are you using an IDE?

https://github.com/google/json_serializable.dart/pull/817/checks?check_run_id=2018018958#step:19:7

looks like you have one fix in the hand-written source and then some updates to the generated source

pratikpparikh commented:
@kevmoo, wow I think I am on the last leg of getting this done.  Can you provide me last set of guidance, what is the _test_yaml level tests?  I ran locally and it seems to be be fine.

C:\opt\development\projects\json-schema\json_serializable.dart\json_serializable>pub run test --run-skipped -t presubmit-only test/ensure_build_test.dart
00:06 +1: All tests passed!

pratikpparikh commented:
@kevmoo I reverted back the test cases with updated formatted message and now the feature update is in clear.

AlBlanc commented:
Will it allow `json_serializable` to support null safety? 🔥

pratikpparikh commented:
@AlBlanc generated code will be in line with the requirement of null-safety

kevmoo commented on json_serializable/lib/src/type_helpers/iterable_helper.dart:
```diff
@@ -70,14 +71,28 @@ class IterableHelper extends TypeHelper<TypeHelperContextWithConfig> {
+        iterableGenericType.isDartCoreInt ||
+        iterableGenericType.isDartCoreNum ||
+        iterableGenericType.isDartCoreString;
+    //
```
delete this

pparikhatliaison replied:
If you delete this it will not cover current null-safe use cases.  The original code did not provide good coverage for existing test cases.  The current test cases are good for new code as well.

pratikpparikh replied:
done

kevmoo commented on json_serializable/test/src/_json_serializable_test_input.dart:
```diff
@@ -398,13 +385,14 @@ enum GoodEnum {
   nullValue
 }
 
+// -
```
oops?

pparikhatliaison replied:
that was mental note for me it can be removed.

pratikpparikh replied:
done, removed it.

kevmoo reviewed (changes requested):
Where are the tests that exercise your fixes? I see you changed a bunch of code, but I don't see how you addressed 803 or 806. I'd expect to see *new* test coverage for these.

pparikhatliaison commented:
> Where are the tests that exercise your fixes? I see you changed a bunch of code, but I don't see how you addressed 803 or 806. I'd expect to see _new_ test coverage for these.

You are correct.  Let review the #803 related code, I think this might only address #806.   I will update you tonight.

pratikpparikh commented:
> Where are the tests that exercise your fixes? I see you changed a bunch of code, but I don't see how you addressed 803 or 806. I'd expect to see _new_ test coverage for these.

the existing tests cover the code.  No additional scenarios are created.  For some reason I had thought that I had already fixed 803 but you are correct not reviewing it back it does not seems I have anything related to it.  So I removed that my title.  Sorry for confusion.

pratikpparikh commented:
@kevmoo can you please find cycles to provide feedback on this PR?

kevmoo commented:
@pratikpparikh – where is the test that changed due to this update?

We should have a case that was broken before and now passes.

Heads up: I'm worried about the amount of churn here, but I'll dig in.

pratikpparikh commented:
> @pratikpparikh – where is the test that changed due to this update?
> 
> We should have a case that was broken before and now passes.
> 
> Heads up: I'm worried about the amount of churn here, but I'll dig in.

I am not sure that is required if the prior fix was just trying to satisfy null safe but with the latest fix actually make it operational.  The test that really broke were the output comparison test.  Review the once in https://github.com/google/json_serializable.dart/blob/master/json_serializable/test/json_serializable_test.dart which depend on src.  So those were the main test needed adjustment.   Also I will tell you this code was not the easiest to navigate so he churn was unavoidable in addition this was my first interaction with all that was required for this library contribution.

I will tell you some of my libraries (around 5 or 6 o them) that did not work prior are not operational due to the fix that is in 206.  Let me know what you suggest to move this forward.

pratikpparikh commented:
@kevmoo do you need me add more test?

kevmoo commented:
Sorry for the lag here.

I'm waiting for `source_gen` to get landed – which itself is blocked.

Then I'm going to migrate this package to null safety.

Then I'll get back in to this issue.

Sorry. Lots going on right now.

pratikpparikh commented:
> Sorry for the lag here.
> 
> I'm waiting for `source_gen` to get landed – which itself is blocked.
> 
> Then I'm going to migrate this package to null safety.
> 
> Then I'll get back in to this issue.
> 
> Sorry. Lots going on right now.

@kevmoo I understand, the dependencies and blockers are huge.  But won't it make sense to merge and resolve the issue before moving to null safety?  Just seems odd if you do it after updating the code for null safety as the code will be out of sync by then.

kevmoo commented:
@pratikpparikh are you free to chat quick? See your email...

kevmoo commented:
Turns out @pratikpparikh – you were holding it wrong

If you put the right type arguments in for `Converter` -> `implements JsonConverter<List<ListNullSafety>?, List<dynamic>?>`

Everything is fine.

```

@JsonSerializable()
@ListNullSafetyJsonConverter()
class ListNullSafety {
  @JsonKey(defaultValue: 'string', includeIfNull: false)
  String bind;

  @JsonKey(defaultValue: <ListNullSafety>[])
  List<ListNullSafety> nullSafetyChild;

  ListNullSafety(this.bind, this.nullSafetyChild);

  factory ListNullSafety.fromJson(Map<String, dynamic> json) =>
      _$ListNullSafetyFromJson(json);

  Map<String, dynamic> toJson() => _$ListNullSafetyToJson(this);
}

class ListNullSafetyJsonConverter
    implements JsonConverter<List<ListNullSafety>?, List<dynamic>?> {
  const ListNullSafetyJsonConverter();

  @override
  List<ListNullSafety>? fromJson(List<dynamic>? json) => json
      ?.map((e) => ListNullSafety.fromJson(e as Map<String, dynamic>))
      .toList();

  @override
  List<Map<String, dynamic>> toJson(List<ListNullSafety>? object) {
    if (object == null || object.isEmpty) {
      return [<String, dynamic>{}];
    }
    return object.map((e) => e.toJson()).toList();
  }
}
```

pratikpparikh commented:
@kevmoo, humm honestly seems like a workaround then fix.  But I am okay with it, well I now know the code that I use in my platform.

kevmoo commented:
We could add code to verify that the generic arguments are as "wide" as the
function parameters.

On Fri, Mar 19, 2021 at 3:10 PM Pratik Parikh ***@***.***>
wrote:

> @kevmoo <https://github.com/kevmoo>, humm honestly seems like a
> workaround then fix. But I am okay with it, well I now know the code that I
> use in my platform.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/json_serializable.dart/pull/817#issuecomment-803162859>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEFCUAXRNGWQRGB2JHGJTTEPDTXANCNFSM4YNR6BTA>
> .
>

pratikpparikh commented:
@kevmoo let us plan for it, if more people are like me and asking for it.  I will test the newly release version today/tomorrow by updating my libraries.  If I find something wrong, we can look at it that point.
